src/System.Management.Automation/engine/lang/parserutils.cs

```diff
@@ -978,66 +979,101 @@ internal static object ReplaceOperator(ExecutionContext context, IScriptExtent e
                     lvalString = lval?.ToString() ?? string.Empty;
                 }
 
-                return ReplaceOperatorImpl(context, lvalString, rr, substitute);
+                return replacer.Replace(lvalString);
             }
             else
             {
                 List<object> resultList = new List<object>();
                 while (ParserOps.MoveNext(context, errorPosition, list))
                 {
                     string lvalString = PSObject.ToStringParser(context, ParserOps.Current(errorPosition, list));
-                    resultList.Add(ReplaceOperatorImpl(context, lvalString, rr, substitute));
+                    resultList.Add(replacer.Replace(lvalString));
                 }
 
                 return resultList.ToArray();
             }
         }
 
-        /// <summary>
-        /// ReplaceOperator implementation.
-        /// Abstracts away conversion of the optional substitute parameter to either a string or a MatchEvaluator delegate
-        /// and finally returns the result of the final Regex.Replace operation.
-        /// </summary>
-        /// <param name="context">The execution context in which to evaluate the expression.</param>
-        /// <param name="input">The input string.</param>
-        /// <param name="regex">A Regex instance.</param>
-        /// <param name="substitute">The substitute value.</param>
-        /// <returns>The result of the regex.Replace operation.</returns>
-        private static object ReplaceOperatorImpl(ExecutionContext context, string input, Regex regex, object substitute)
+        private struct OperatorReplacer
         {
-            switch (substitute)
+            public static OperatorReplacer Create(ExecutionContext context, Regex regex, object substitute)
             {
-                case string replacementString:
-                    return regex.Replace(input, replacementString);
+                return new OperatorReplacer(context, regex, substitute, replacementString: null, matchEvaluator: null);
+            }
+
+            private readonly ExecutionContext _context;
+            private readonly Regex _regex;
+
+            private string _cachedReplacementString;
+
+            private MatchEvaluator _cachedMatchEvaluator;
+
+            private OperatorReplacer(
+                ExecutionContext context,
+                Regex regex,
+                object substitute,
+                string replacementString,
+                MatchEvaluator matchEvaluator)
```

Comment:
Why not remove the parameters `replacementString` and `matchEvaluator`? They are not used.

Reply:
Compiler complains with https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0171

Reply:
> A constructor in a struct must initialize all fields in the struct.

Just initialize `_cachedReplacementString` and `_cachedMatchEvaluator` to null in the constructor, you don't need those 2 parameters.

Reply:
Done. Thanks!